### PR TITLE
[11.0][FIX] product_supplierinfo_for_customer_sale: Search supplierinfo by product variant first

### PR DESCRIPTION
**Pull request by sergio-teruel**
When you have a product with several variants and supplierinfo for every variant, always set product code from template due to incorrect domain search.

### CODE REVIEW

**Review by rafaelbn — APPROVED** [review]
Functional test in runbot 👍 

Please @sergio-teruel check travist is :red

**Comment by pedrobaeza** [comment]
Little flake8 checks:
```
product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py:42:9: E125 continuation line with same indent as next logical line
product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py:103:62: E202 whitespace before ')'
```

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 1]
```diff
@@ -48,12 +41,46 @@ def product_id_change(self):
...
-                code_id = self.env['product.supplierinfo'].search([
-                    '|', ('product_tmpl_id', '=', product.product_tmpl_id.id),
-                    ('product_id', '=', product.id),
-                    ('supplierinfo_type', '=', 'customer'),
-                    ('name', '=', line.order_id.partner_id.id),
-                ], limit=1)
-                if code_id and code_id.min_qty:
-                    line.product_uom_qty = code_id.min_qty
+                supplierinfo = self.get_supplierinfo(line)
+                line.product_uom_qty = supplierinfo and supplierinfo.min_qty
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: If no `supplierinfo.min_qty`, you shouldn't set it IMO

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 2]
```diff
@@ -48,12 +41,46 @@ def product_id_change(self):
...
+        Search supplierinfo only by product template if it has not variant for
+        performance.
+        """
+        if line.product_id.product_variant_count > 1:
+            return self._search_supplierinfo_by_variant(line)
+        else:
+            return self._search_supplierinfo_by_template(line)
+
+    def _supplierinfo_domain(self, line):
+        """
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: This docstring can be put in one line.

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 3]
```diff
@@ -48,12 +41,46 @@ def product_id_change(self):
...
+                line.product_uom_qty = supplierinfo and supplierinfo.min_qty
         return result
+
+    def get_supplierinfo(self, line):
+        """
+        Search supplierinfo only by product template if it has not variant for
+        performance.
+        """
+        if line.product_id.product_variant_count > 1:
+            return self._search_supplierinfo_by_variant(line)
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: You should always search by this condition, as there could be in the past several variants that are now deactivated by any reason and their supplierinfo records stay there yet, and you are returning the incorrect one.

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 4]
```diff
@@ -48,12 +41,46 @@ def product_id_change(self):
...
+        """
+        Common domain for product templates and variants
+        """
+        return [
+            ('supplierinfo_type', '=', 'customer'),
+            ('name', '=', line.order_partner_id.id),
+        ]
+
+    def _search_supplierinfo_by_template(self, line):
+        """
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: This docstring can be put in one line.

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 5]
```diff
@@ -48,12 +41,46 @@ def product_id_change(self):
...
+        Search only supplierinfo by template
+        """
+        return self.env['product.supplierinfo'].search(
+            self._supplierinfo_domain(line) +
+            [('product_tmpl_id', '=', line.product_id.product_tmpl_id.id)],
+            limit=1)
+
+    def _search_supplierinfo_by_variant(self, line):
+        """
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py:57]: This docstring can be put in one line.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Why not using `_select_seller` method passing as context `supplierinfo_type='customer'` and the rest will work properly?

**Comment by sergio-teruel** [comment]
@pedrobaeza ¿ Are yo referrer to this method https://github.com/odoo/odoo/blob/e3f5f175614d0820af951a8b6c99982638ab9c87/addons/product/models/product.py#L503?

**Comment by rafaelbn** [comment]
@pedrobaeza please could you review comments?

File: product_supplierinfo_for_customer_sale/README.rst [hunk 6]
```diff
@@ -35,30 +35,30 @@ To use this module, you need:
 - Go to product and configure *Partner product name* and *Partner product code*
   for each selected customer.
 
-.. figure:: https://raw.githubusercontent.com/OCA/sale-workflow/11.0/product_supplierinfo_for_customer_sale/static/description/configuration_customer.png
+.. figure:: static/description/configuration_customer.png
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/README.rst]: It can be that you have an old README generator tool? This shouldn't be modified

Comment (pedrobaeza) [reply product_supplierinfo_for_customer_sale/README.rst]: And any way, why regenerating it?

**Comment by pedrobaeza** [comment]
@sergio-teruel did you try to use the method `_select_seller`?

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 7]
```diff
@@ -8,26 +8,19 @@
...
-                '|', ('product_tmpl_id', '=', product.product_tmpl_id.id),
-                ('product_id', '=', product.id),
-                ('supplierinfo_type', '=', 'customer'),
-                ('name', '=', line.order_id.partner_id.id),
-            ], limit=1)
-            line.product_customer_code = code_id.product_code or ''
-
     product_customer_code = fields.Char(
-        compute='_compute_get_product_customer_code',
+        compute='_compute_product_customer_code',
```
Comment (Swapnesh-SerpentCS) [inline product_supplierinfo_for_customer_sale/models/sale.py:12]: I don't think this should be on existing module

**Comment by rafaelbn** [comment]
The migration of this module to v12 is here : https://github.com/OCA/sale-workflow/pull/841

**Comment by rafaelbn** [comment]
@pedrobaeza please could you unblock here?

File: product_supplierinfo_for_customer_sale/models/sale.py (hunk 3 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer_sale/models/sale.py]: @sergio-teruel what about this comment?

**Comment by mreficent** [comment]
conflicts

File: product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py [hunk 8]
```diff
@@ -68,3 +90,46 @@ def test_product_supplierinfo_for_customer_sale(self):
...
+        line = self.env['sale.order.line'].create({
+            'product_id': self.product_variant_2.id,
+            'order_id': so.id,
+        })
+        line.product_id_change()
+        self.assertEqual(
+            line.product_customer_code, supplierinfo.product_code,
+            "Error: Customer product code was not passed to sale order line")
+        # Test with product without variants
+        so2 = self.env['sale.order'].create({
```
Comment (MiquelRForgeFlow) [inline product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py:124]:
```suggestion
        self.env['sale.order'].create({
```

Comment (sergio-teruel) [reply product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py:124]: copy paste error... line2 linked to so2

File: product_supplierinfo_for_customer_sale/README.rst (hunk 6 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer_sale/README.rst]: This is still wrong

File: product_supplierinfo_for_customer_sale/__manifest__.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
     "summary": "Loads in every sale order line the customer code defined "
                "in the product",
     "author": "Agile Business Group,Vauxoo,Odoo Community Association (OCA)",
-    "website": "https://github.com/OCA/sale-workflow/tree/11.0/product_supplierinfo_for_customer_sale", # noqa
+    "website": "https://github.com/OCA/sale-workflow/tree/11.0/product_supplierinfo_for_customer_sale",  # noqa
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/__manifest__.py]: Please put short URL (without full path)

File: product_supplierinfo_for_customer_sale/models/sale.py (hunk 3 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer_sale/models/sale.py]: This comment is still pending

File: product_supplierinfo_for_customer_sale/static/description/index.html [hunk 10]
```diff
@@ -3,7 +3,7 @@
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
 <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
-<meta name="generator" content="Docutils 0.15.1: http://docutils.sourceforge.net/" />
+<meta name="generator" content="Docutils 0.15.2: http://docutils.sourceforge.net/" />
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/static/description/index.html]: Incorrect docutils version or this one is outdated?

**Comment by pedrobaeza** [comment]
@sergio-teruel please finish this. We meet tomorrow if you prefer for details.

File: product_supplierinfo_for_customer_sale/models/sale.py (hunk 3 above)
Comment (sergio-teruel) [reply product_supplierinfo_for_customer_sale/models/sale.py]: Ok, in which one are you prefer?

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 11]
```diff
@@ -48,12 +41,51 @@ def product_id_change(self):
...
-                    ('name', '=', line.order_id.partner_id.id),
-                ], limit=1)
-                if code_id and code_id.min_qty:
-                    line.product_uom_qty = code_id.min_qty
+                supplierinfo = self.get_supplierinfo(line)
+                if supplierinfo and supplierinfo.min_qty:
+                    line.product_uom_qty = supplierinfo.min_qty
         return result
+
+    def get_supplierinfo(self, line):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: Shouldn't this be called `get_customer_supplierinfo` for clarity?

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 12]
```diff
@@ -48,12 +41,51 @@ def product_id_change(self):
...
+        return self._search_supplierinfo_by_variant(line)
+
+    def _supplierinfo_domain(self, line):
+        """Common domain for product templates and variants"""
+        return [
+            ('supplierinfo_type', '=', 'customer'),
+            ('name', '=', line.order_partner_id.id),
+        ]
+
+    def _search_supplierinfo_by_template(self, line):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: Do we still need this method? I don't think so.

File: product_supplierinfo_for_customer_sale/models/sale.py [hunk 13]
```diff
@@ -48,12 +41,51 @@ def product_id_change(self):
...
+                if supplierinfo and supplierinfo.min_qty:
+                    line.product_uom_qty = supplierinfo.min_qty
         return result
+
+    def get_supplierinfo(self, line):
+        """
+        Search supplierinfo only by product template if it has not variant for
+        performance.
+        """
+        return self._search_supplierinfo_by_variant(line)
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/models/sale.py]: Better to directly put here the code of `_search_supplierinfo_by_variant`

File: product_supplierinfo_for_customer_sale/readme/USAGE.rst [hunk 14]
```diff
@@ -0,0 +1,31 @@
+To use this module, you need:
+
+- Go to product and configure *Partner product name* and *Partner product code*
+  for each selected customer.
+
+.. image:: ../static/description/configuration_customer.png
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/readme/USAGE.rst]: Why this is changed from `figure` to `image`?

File: product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py [hunk 15]
```diff
@@ -51,7 +73,7 @@ def _create_pricelist(self, name, product):
             })],
         })
 
-    def test_product_supplierinfo_for_customer_sale(self):
+    def _test_product_supplierinfo_for_customer_sale(self):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py]: This test has been disabled

File: product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py [hunk 16]
```diff
@@ -68,3 +90,63 @@ def test_product_supplierinfo_for_customer_sale(self):
         self.assertEqual(
             line.product_uom_qty, self.supplierinfo.min_qty,
             "Error: Min qty was not passed to the sale order line")
+
+    def _test_product_supplierinfo_for_customer_sale_variant(self):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py]: This test is disabled

File: product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py [hunk 17]
```diff
@@ -68,3 +90,63 @@ def test_product_supplierinfo_for_customer_sale(self):
...
+        line = self.env['sale.order.line'].create({
+            'product_id': self.product_variant_1.id,
+            'order_id': so.id,
+        })
+        line.product_id_change()
+        self.assertEqual(
+            line.product_customer_code, self.supplierinfo.product_code,
+            "Error: Customer product code was not passed to sale order line")
+
+    def _test_product_supplierinfo_for_customer_sale_template(self):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer_sale/tests/test_product_supplierinfo_for_customer_sale.py]: This test is disabled

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
Upsssss!!! changes done!!

**Comment by pedrobaeza** [comment]
I'm not seeing several of my comments attended nor answered

**Comment by sergio-teruel** [comment]
REady

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**